src/Microsoft.AspNetCore.Http/StreamPipeWriter.cs

```diff
@@ -0,0 +1,313 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Http
+{
+    /// <summary>
+    /// Implements PipeWriter using a underlying stream. 
+    /// </summary>
+    public class StreamPipeWriter : PipeWriter, IDisposable
+    {
+        private readonly int _minimumSegmentSize;
+        private readonly Stream _writingStream;
+        private int _bytesWritten;
+
+        private List<CompletedBuffer> _completedSegments;
+        private Memory<byte> _currentSegment;
+        private IMemoryOwner<byte> _currentSegmentOwner;
+        private int _position;
+
+        private CancellationTokenSource _internalTokenSource;
+        private bool _isCompleted;
+        private ExceptionDispatchInfo _exceptionInfo;
+
+        private CancellationTokenSource InternalTokenSource
+        {
+            get
+            {
+                if (_internalTokenSource == null)
+                {
+                    _internalTokenSource = new CancellationTokenSource();
+                }
+                return _internalTokenSource;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new StreamPipeWrapper 
+        /// </summary>
+        /// <param name="writingStream">The stream to write to</param>
+        public StreamPipeWriter(Stream writingStream) : this(writingStream, 4096)
+        {
+        }
+
+        public StreamPipeWriter(Stream writingStream, int minimumSegmentSize, MemoryPool<byte> pool = null)
+        {
+            _minimumSegmentSize = minimumSegmentSize;
+            _writingStream = writingStream;
+        }
+
+        /// <inheritdoc />
+        public override void Advance(int count)
+        {
+            if (_currentSegment.IsEmpty) // TODO confirm this
+            {
+                throw new InvalidOperationException("No writing operation. Make sure GetMemory() was called.");
+            }
+
+            if (count >= 0)
+            {
+                if (_currentSegment.Length < _position + count)
+                {
+                    throw new InvalidOperationException("Can't advance past buffer size.");
+                }
+                _bytesWritten += count;
+                _position += count;
+            }
+        }
+
+        /// <inheritdoc />
+        public override Memory<byte> GetMemory(int sizeHint = 0)
+        {
+            EnsureCapacity(sizeHint);
+
+            return _currentSegment;
+        }
+
+        /// <inheritdoc />
+        public override Span<byte> GetSpan(int sizeHint = 0)
+        {
+            EnsureCapacity(sizeHint);
+
+            return _currentSegment.Slice(_position).Span;
```

Comment:
I believe getting a span then slicing is faster then slicing memory

Reply:
Yep